Source/Core/Core/IOS/USB/Bluetooth/BTReal.cpp

```diff
@@ -60,259 +56,260 @@ BluetoothRealDevice::BluetoothRealDevice(EmulationKernel& ios, const std::string
     {
-      CriticalAlertFmtT(
-          "Could not find any usable Bluetooth USB adapter for Bluetooth Passthrough.\n\n"
-          "The emulated console will now stop.");
+      m_acl_endpoint = std::move(cmd);
+      TryToFillACLEndpoint();
+      return std::nullopt;
     }
-    else
+
+    auto& memory = GetSystem().GetMemory();
+    const auto payload = memory.GetSpanForAddress(cmd->data_address).first(cmd->length);
+    m_lib_usb_bt_adapter->ScheduleBulkTransfer(cmd->endpoint, payload, GetTargetTime());
+    break;
+  }
+  // Incoming HCI events
+  case USB::IOCTLV_USBV0_INTRMSG:
+  {
+    auto cmd = std::make_unique<USB::V0IntrMessage>(GetEmulationKernel(), request);
+    if (cmd->endpoint == HCI_EVENT)
     {
-      CriticalAlertFmtT(
-          "Could not find any usable Bluetooth USB adapter for Bluetooth Passthrough.\n"
-          "The following error occurred when Dolphin tried to use an adapter:\n{0}\n\n"
-          "The emulated console will now stop.",
-          m_last_open_error);
+      m_hci_endpoint = std::move(cmd);
+      TryToFillHCIEndpoint();
+      return std::nullopt;
     }
-    Core::QueueHostJob(&Core::Stop);
-    return IPCReply(IPC_ENOENT);
+
+    ERROR_LOG_FMT(IOS_WIIMOTE, "IOCTLV_USBV0_INTRMSG: Unknown endpoint: 0x{:02x}", cmd->endpoint);
+  }
+  default:
+    ERROR_LOG_FMT(IOS_WIIMOTE, "IOCtlV: Unknown request: 0x{:08x}", request.request);
+    break;
   }
 
-  return Device::Open(request);
+  return IPCReply{IPC_SUCCESS};
 }
 
-std::optional<IPCReply> BluetoothRealDevice::Close(u32 fd)
+void BluetoothRealDevice::Update()
+{
+  if (m_hci_endpoint != nullptr)
+    TryToFillHCIEndpoint();
+
+  if (m_acl_endpoint != nullptr)
+    TryToFillACLEndpoint();
+}
+
+void BluetoothRealDevice::TryToFillHCIEndpoint()
 {
-  if (m_handle)
+  if (m_sync_button_state == SyncButtonState::Pressed)
+  {
+    Core::DisplayMessage("Scanning for Wii Remotes", 2000);
+    FakeSyncButtonPressedEvent(*m_hci_endpoint);
+  }
+  else if (m_sync_button_state == SyncButtonState::LongPressed)
+  {
+    Core::DisplayMessage("Reset saved Wii Remote pairings", 2000);
+    FakeSyncButtonHeldEvent(*m_hci_endpoint);
+  }
+  else if (!m_fake_replies.empty())
   {
-    const int ret = libusb_release_interface(m_handle, 0);
-    if (ret != LIBUSB_SUCCESS)
-      WARN_LOG_FMT(IOS_WIIMOTE, "libusb_release_interface failed: {}", LibusbUtils::ErrorWrap(ret));
-    libusb_close(m_handle);
-    libusb_unref_device(m_device);
-    m_handle = nullptr;
+    std::invoke(m_fake_replies.front(), *m_hci_endpoint);
+    m_fake_replies.pop();
   }
+  else
+  {
+    const auto buffer = ProcessHCIEvent(m_lib_usb_bt_adapter->ReceiveHCIEvent());
+    if (buffer.empty())
+      return;
 
-  return Device::Close(fd);
+    assert(buffer.size() <= m_hci_endpoint->length);
+
+    m_hci_endpoint->FillBuffer(buffer.data(), buffer.size());
+    GetEmulationKernel().EnqueueIPCReply(m_hci_endpoint->ios_request, s32(buffer.size()));
+  }
+
+  m_hci_endpoint.reset();
```

Comment:
Should this be wrapped in a scoped object so any early returns still trigger it?